File: packages/flags-python/src/signakit_flags/audience_matcher.py
"""Audience matching for evaluating user attributes against rule conditions."""

from __future__ import annotations

import logging

from .types import (
    AudienceCondition,
    AudienceMatchType,
    ConfigRuleAudience,
    UserAttributes,
)

_logger = logging.getLogger(__name__)


def matches_condition(
    condition: AudienceCondition, attributes: UserAttributes | None
) -> bool:
    """Return ``True`` if a single condition matches the given attributes."""
    if attributes is None:
        return False

    if condition.attribute not in attributes:
        return False

    user_value = attributes[condition.attribute]
    op = condition.operator
    value = condition.value

    # Treat bool specially: in Python, bool is a subclass of int, so we must
    # avoid accidentally matching it under numeric comparisons.
    if op == "equals":
        return user_value == value
    if op == "not_equals":
        return user_value != value

    if op in ("greater_than", "less_than", "greater_than_or_equals", "less_than_or_equals"):
        if (
            isinstance(user_value, (int, float))
            and not isinstance(user_value, bool)
            and isinstance(value, (int, float))
            and not isinstance(value, bool)
        ):
            if op == "greater_than":
                return user_value > value
            if op == "less_than":
                return user_value < value
            if op == "greater_than_or_equals":
                return user_value >= value
            return user_value <= value
        return False

    if op == "in":
        if isinstance(value, list):
            return user_value in value
        return False

    if op == "not_in":
        if isinstance(value, list):
            return user_value not in value
        return True

    if op == "contains":
        if isinstance(user_value, str) and isinstance(value, str):
            return value in user_value
        if isinstance(user_value, list) and isinstance(value, str):
            return value in user_value
        return False

    if op == "not_contains":
        if isinstance(user_value, str) and isinstance(value, str):
            return value not in user_value
        if isinstance(user_value, list) and isinstance(value, str):
            return value not in user_value
        return True

    _logger.warning("[SignaKit] Unknown operator: %s", op)
    return False


def matches_audience(
    audience: ConfigRuleAudience, attributes: UserAttributes | None
) -> bool:
    """All conditions must match for an audience to match."""
    return all(matches_condition(c, attributes) for c in audience.conditions)


def matches_audiences(
    audiences: list[ConfigRuleAudience] | None,
    match_type: AudienceMatchType | None,
    attributes: UserAttributes | None,
) -> bool:
    """Evaluate multiple audiences with ``any``/``all`` semantics.

    No audiences ⇒ rule matches all users (parity with the JS SDK).
    """
    if not audiences:
        return True

    if match_type == "any":
        return any(matches_audience(a, attributes) for a in audiences)
    return all(matches_audience(a, attributes) for a in audiences)

File: packages/flags-react-native/src/audience-matcher.ts
/**
 * Audience matching logic for evaluating user attributes against audience conditions.
 */

import type {
  AudienceCondition,
  ConfigRuleAudience,
  AudienceMatchType,
  UserAttributes,
} from './types'

export function matchesCondition(
  condition: AudienceCondition,
  attributes: UserAttributes | undefined
): boolean {
  if (!attributes) return false

  const userValue = attributes[condition.attribute]
  if (userValue === undefined) return false

  const { operator, value } = condition

  switch (operator) {
    case 'equals':
      return userValue === value

    case 'not_equals':
      return userValue !== value

    case 'greater_than':
      if (typeof userValue === 'number' && typeof value === 'number') {
        return userValue > value
      }
      return false

    case 'less_than':
      if (typeof userValue === 'number' && typeof value === 'number') {
        return userValue < value
      }
      return false

    case 'greater_than_or_equals':
      if (typeof userValue === 'number' && typeof value === 'number') {
        return userValue >= value
      }
      return false

    case 'less_than_or_equals':
      if (typeof userValue === 'number' && typeof value === 'number') {
        return userValue <= value
      }
      return false

    case 'in':
      if (Array.isArray(value)) {
        return (value as unknown[]).includes(userValue)
      }
      return false

    case 'not_in':
      if (Array.isArray(value)) {
        return !(value as unknown[]).includes(userValue)
      }
      return true

    case 'contains':
      if (typeof userValue === 'string' && typeof value === 'string') {
        return userValue.includes(value)
      }
      if (Array.isArray(userValue) && typeof value === 'string') {
        return userValue.includes(value)
      }
      return false

    case 'not_contains':
      if (typeof userValue === 'string' && typeof value === 'string') {
        return !userValue.includes(value)
      }
      if (Array.isArray(userValue) && typeof value === 'string') {
        return !userValue.includes(value)
      }
      return true

    default:
      console.warn(`[SignaKit] Unknown operator: ${operator}`)
      return false
  }
}

export function matchesAudience(
  audience: ConfigRuleAudience,
  attributes: UserAttributes | undefined
): boolean {
  return audience.conditions.every((condition) => matchesCondition(condition, attributes))
}

export function matchesAudiences(
  audiences: ConfigRuleAudience[] | undefined,
  matchType: AudienceMatchType | undefined,
  attributes: UserAttributes | undefined
): boolean {
  if (!audiences || audiences.length === 0) return true

  if (matchType === 'any') {
    return audiences.some((audience) => matchesAudience(audience, attributes))
  }
  return audiences.every((audience) => matchesAudience(audience, attributes))
}
